File: sdk-ts/src/flowsdk/v1beta2/input.ts
import { create, type DescField, type DescMessage, fromJson, getOption, hasOption, type Message, type Registry } from "@bufbuild/protobuf";

import { type Input, Input_BoolBindingSchema, Input_BytesBindingSchema, Input_DoubleBindingSchema, Input_FloatBindingSchema, Input_Int32BindingSchema, Input_Int64BindingSchema, Input_ListBindingSchema, Input_MapBindingSchema, Input_StringBindingSchema, Input_Uint32BindingSchema, Input_Uint64BindingSchema } from "../../proto/dtkt/flow/v1beta2/spec_pb";
import { field as fieldExt, type FieldElement, FieldElementSchema } from "../../proto/dtkt/protoform/v1beta1/protoform_pb";

export type InputBindingResult = {
  /**
   * The protoform FieldElement driving the widget. For wrapper
   * bindings (scalar / list / map) it's the annotation lifted from
   * the binding's `value` field. For typed-message bindings it's a
   * synthetic element holding the input id as title -- per-field
   * annotations on the typed message drive the actual widgets.
   */
  element: FieldElement;
  /**
   * The binding proto, pre-populated from the input's default when
   * set. For scalar / list / map inputs it's an Input_*Binding
   * wrapper. For Message inputs it's a fresh instance of the
   * declared FQN type (its own fields drive the form).
   */
  binding: Message;
  /**
   * Descriptor for the returned binding -- consumers use this to
   * `reflect(desc, msg)` for protoform rendering or `anyPack(desc,
   * msg)` for wire encoding.
   */
  desc: DescMessage;
};

/**
 * Returns the protoform FieldElement and binding proto for a flow
 * Input. Tuiform / protoformsdk consumes the binding's
 * protoform-annotated fields and renders the matching widgets; the
 * binding is pre-populated from the input's typed default when set.
 *
 * For scalar / list / map inputs the binding is one of Input_*Binding
 * (a wrapper whose `value` field carries the FieldElement annotation).
 * For Message inputs the registry looks up the declared FQN and the
 * returned binding is a fresh instance of *that* message -- its own
 * protoform-annotated fields drive the form. The Message default
 * (a Struct) populates the typed message via a protojson round-trip.
 *
 * Mirrors `GetInputBinding` in `sdk-go/flowsdk/v1beta2/input.go`.
 *
 * Returns `undefined` when the input's type is unrecognised (oneof
 * unset), the message FQN can't be resolved by the registry, or the
 * default can't be coerced into the resolved type.
 */
export function getInputBinding(input: Input, registry: Registry): InputBindingResult | undefined {
  const id = input.id;

  switch (input.type.case) {
    case "bool": {
      const binding = create(Input_BoolBindingSchema);
      if (input.type.value.default !== undefined) {
        binding.value = input.type.value.default;
      }
      return wrapperResult(binding, Input_BoolBindingSchema, id);
    }
    case "bytes": {
      const binding = create(Input_BytesBindingSchema);
      if (input.type.value.default !== undefined) {
        binding.value = input.type.value.default;
      }
      return wrapperResult(binding, Input_BytesBindingSchema, id);
    }
    case "double": {
      const binding = create(Input_DoubleBindingSchema);
      if (input.type.value.default !== undefined) {
        binding.value = input.type.value.default;
      }
      return wrapperResult(binding, Input_DoubleBindingSchema, id);
    }
    case "float": {
      const binding = create(Input_FloatBindingSchema);
      if (input.type.value.default !== undefined) {
        binding.value = input.type.value.default;
      }
      return wrapperResult(binding, Input_FloatBindingSchema, id);
    }
    case "int64": {
      const binding = create(Input_Int64BindingSchema);
      if (input.type.value.default !== undefined) {
        binding.value = input.type.value.default;
      }
      return wrapperResult(binding, Input_Int64BindingSchema, id);
    }
    case "uint64": {
      const binding = create(Input_Uint64BindingSchema);
      if (input.type.value.default !== undefined) {
        binding.value = input.type.value.default;
      }
      return wrapperResult(binding, Input_Uint64BindingSchema, id);
    }
    case "int32": {
      const binding = create(Input_Int32BindingSchema);
      if (input.type.value.default !== undefined) {
        binding.value = input.type.value.default;
      }
      return wrapperResult(binding, Input_Int32BindingSchema, id);
    }
    case "uint32": {
      const binding = create(Input_Uint32BindingSchema);
      if (input.type.value.default !== undefined) {
        binding.value = input.type.value.default;
      }
      return wrapperResult(binding, Input_Uint32BindingSchema, id);
    }
    case "string": {
      const binding = create(Input_StringBindingSchema);
      if (input.type.value.default !== undefined) {
        binding.value = input.type.value.default;
      }
      return wrapperResult(binding, Input_StringBindingSchema, id);
    }
    case "list": {
      const binding = create(Input_ListBindingSchema);
      if (input.type.value.default !== undefined) {
        binding.value = input.type.value.default;
      }
      return wrapperResult(binding, Input_ListBindingSchema, id);
    }
    case "map": {
      const binding = create(Input_MapBindingSchema);
      if (input.type.value.default !== undefined) {
        binding.value = input.type.value.default;
      }
      return wrapperResult(binding, Input_MapBindingSchema, id);
    }
    case "message": {
      const desc = registry.getMessage(input.type.value.type);
      if (!desc) {
        return undefined;
      }
      let binding: Message;
      if (input.type.value.default !== undefined) {
        // Lift the Struct-shaped default (typed as JsonObject in TS
        // proto due to bufbuild's Struct -> JSON specialization) into
        // the declared FQN via protojson. Mirrors
        // common.ProtoOptions.AsMessage on the Go side.
        try {
          binding = fromJson(desc, input.type.value.default, { registry });
        } catch {
          return undefined;
        }
      } else {
        binding = create(desc);
      }
      return {
        element: create(FieldElementSchema, { title: id }),
        binding,
        desc,
      };
    }
  }

  return undefined;
}

/**
 * Builds a wrapper-binding result by lifting the FieldElement
 * annotation off the binding's `value` field. The annotation drives
 * widget choice (Confirm / Input single-line / Input multiline /
 * etc.); the input id overrides the element title since flow.Input
 * has no title field of its own.
 */
function wrapperResult(binding: Message, schema: DescMessage, id: string): InputBindingResult {
  const valueField: DescField | undefined = schema.field.value;
  let element: FieldElement;
  if (valueField && hasOption(valueField, fieldExt)) {
    element = getOption(valueField, fieldExt);
  } else {
    element = create(FieldElementSchema);
  }
  element.title = id;
  return { element, binding, desc: schema };
}

File: sdk-go/flowsdk/v1beta2/input.go
package v1beta2

import (
	"github.com/datakit-dev/dtkt-sdk/sdk-go/common"
	flowv1beta2 "github.com/datakit-dev/dtkt-sdk/sdk-go/proto/dtkt/flow/v1beta2"
	protoformv1beta1 "github.com/datakit-dev/dtkt-sdk/sdk-go/proto/dtkt/protoform/v1beta1"
	form "github.com/datakit-dev/dtkt-sdk/sdk-go/protoformsdk/v1beta1"
	"github.com/jhump/protoreflect/v2/protoresolve"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/reflect/protoreflect"
)

// GetInputBinding returns the protoform FieldElement and binding proto
// for a flow Input. Tuiform / protoformsdk consumes the binding's
// protoform-annotated fields and renders the matching widgets; the
// binding has its value pre-populated from the input's typed default
// when set.
//
// For scalar / list / map inputs the binding is one of Input_*Binding
// (a wrapper whose `value` field carries the FieldElement annotation).
// For Input_Message inputs the resolver looks up the declared FQN and
// the returned binding is a fresh instance of *that* message -- its
// own protoform-annotated fields drive the form. The Message default
// (a Struct) populates the typed message via the same protojson
// round-trip the runtime uses (see common.ProtoOptions.AsMessage).
//
// resolver is required only for Input_Message; other types ignore it.
//
// Returns (nil, nil, false) when the input's type is unrecognised
// (oneof unset), the message FQN can't be resolved by the resolver,
// or the default can't be coerced into the resolved type.
func GetInputBinding(input *flowv1beta2.Input, resolver protoresolve.SerializationResolver) (*protoformv1beta1.FieldElement, proto.Message, bool) {
	var binding proto.Message
	switch t := input.GetType().(type) {
	case *flowv1beta2.Input_Bool:
		b := &flowv1beta2.Input_BoolBinding{}
		if t.Bool.HasDefault() {
			b.SetValue(t.Bool.GetDefault())
		}
		binding = b
	case *flowv1beta2.Input_Bytes:
		b := &flowv1beta2.Input_BytesBinding{}
		if t.Bytes.HasDefault() {
			b.SetValue(t.Bytes.GetDefault())
		}
		binding = b
	case *flowv1beta2.Input_Double:
		b := &flowv1beta2.Input_DoubleBinding{}
		if t.Double.HasDefault() {
			b.SetValue(t.Double.GetDefault())
		}
		binding = b
	case *flowv1beta2.Input_Float:
		b := &flowv1beta2.Input_FloatBinding{}
		if t.Float.HasDefault() {
			b.SetValue(t.Float.GetDefault())
		}
		binding = b
	case *flowv1beta2.Input_Int64:
		b := &flowv1beta2.Input_Int64Binding{}
		if t.Int64.HasDefault() {
			b.SetValue(t.Int64.GetDefault())
		}
		binding = b
	case *flowv1beta2.Input_Uint64:
		b := &flowv1beta2.Input_Uint64Binding{}
		if t.Uint64.HasDefault() {
			b.SetValue(t.Uint64.GetDefault())
		}
		binding = b
	case *flowv1beta2.Input_Int32:
		b := &flowv1beta2.Input_Int32Binding{}
		if t.Int32.HasDefault() {
			b.SetValue(t.Int32.GetDefault())
		}
		binding = b
	case *flowv1beta2.Input_Uint32:
		b := &flowv1beta2.Input_Uint32Binding{}
		if t.Uint32.HasDefault() {
			b.SetValue(t.Uint32.GetDefault())
		}
		binding = b
	case *flowv1beta2.Input_String_:
		b := &flowv1beta2.Input_StringBinding{}
		if t.String_.HasDefault() {
			b.SetValue(t.String_.GetDefault())
		}
		binding = b
	case *flowv1beta2.Input_List:
		b := &flowv1beta2.Input_ListBinding{}
		if t.List.HasDefault() {
			b.SetValue(t.List.GetDefault())
		}
		binding = b
	case *flowv1beta2.Input_Map:
		b := &flowv1beta2.Input_MapBinding{}
		if t.Map.HasDefault() {
			b.SetValue(t.Map.GetDefault())
		}
		binding = b
	case *flowv1beta2.Input_Message:
		msgType, err := resolver.FindMessageByName(protoreflect.FullName(t.Message.GetType()))
		if err != nil {
			return nil, nil, false
		}
		if t.Message.HasDefault() {
			msg, err := (common.ProtoOptions{Resolver: resolver}).AsMessage(t.Message.GetDefault(), msgType)
			if err != nil {
				return nil, nil, false
			}
			binding = msg
		} else {
			binding = msgType.New().Interface()
		}
	}

	if binding == nil {
		return nil, nil, false
	}

	// For wrapper bindings (scalar / list / map), the FieldElement
	// lives on the binding's `value` field and is the single widget
	// hint. Surface the input id as its title since flow.Input has
	// no title field.
	//
	// For typed-message bindings there is no `value` field -- the
	// returned proto IS the form, and each of its fields carries its
	// own FieldElement annotation. We return a synthetic top-level
	// element holding the id; the caller iterates the field group
	// and the per-field annotations drive each widget.
	if desc := binding.ProtoReflect().Descriptor().Fields().ByName("value"); desc != nil {
		if elem, ok := form.GetFieldElement(desc); ok {
			title := input.GetId()
			elem.Title = &title
			return elem, binding, true
		}
	}

	title := input.GetId()
	return &protoformv1beta1.FieldElement{Title: &title}, binding, true
}
